app/helpers/sessions_helper.rb

```diff
@@ -2,21 +2,53 @@ module SessionsHelper
   # Logs in the given user
   def log_in user
     session[:user_id] = user.id
+    session[:session_token] = user.session_token
   end
 
   # Returns the current logged-in user (if any)
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    @current_user ||= user_from_session || user_from_cookies
   end
 
   # Returns true if the user is logged in, false otherwise
   def logged_in?
-    current_user.present?
+    !current_user.nil?
   end
 
   # Logs out the current user.
   def log_out
+    forget current_user
     reset_session
     @current_user = nil
   end
+
+  def remember_cookies user
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
+  def forget user
+    user.forget
+    cookies.delete(:user_id)
+    cookies.delete(:remember_token)
+  end
+
+  private
+  def user_from_session
+    return unless (user_id = session[:user_id])
+
+    user = User.find_by id: user_id
+    user if user&.session_token == session[:session_token]
```

Comment:
```suggestion
    user if user&.authenticated?(session[:session_token])
```
Chỗ này check authen session[:session_token]